README.md
ABOUT THIS MODULE
-----------------
The NYPL Advocacy module was written in 2012 by Dillon Savage and Clint Newsom to facilitate the New York Public Library's efforts to reduce the City's annual cut to the Library's budget. The module allows website users to write to the mayor and members of the City Council with a customizable message of support for the Library, and/or to suggest that others do the same via a "Tell a Friend" form. It also allows Library staff with the correct permissions to enter information about physical letters sent from library branches, and to export physical and electronic letter submission data.

INSTALLATION
------------
NYPL Advocacy can be installed and enabled in the manner described here: http://drupal.org/node/70151.

When it's enabled for the first time, the module creates two content types, City Council member and Advocacy Landing Page. It creates a node for each Council member (N.B. as of 2012) and two Webform nodes, "Advocacy Form" and "Tell a Friend". It creates the nypl_advocacy database table, which stores the physical letter data. Finally, the module specifies the pages 'speakout/thankyou' (where the user is redirected after filling out the City Council form) and 'speakout/toolbox' (where the user can download icons for use in social media).

The Webform nodes created by NYPL Advocacy's installation script are not meant to be used in production. Before deploying them to a production setting, a site administrator should make any campaign-specific modifications that might be necessary. Perhaps most importantly, in order for City Council form submissions to arrive at their destinations, the City Council Webform must contain an email setting whose "email to" value pulls from the CC form's hidden "email to" field.

CONFIGURATION
-------------
When the module is enabled, a number of configuration options are available at admin/settings/nypl_advocacy. Additionally, staff can enter physical letter statistics and export data by clicking the appropriate tabs on this page.

DEPENDENCIES
------------
NYPL Advocacy depends on functionality provided by the Webform, Locations (NYPL custom), and GMap Drupal modules.

To match user-entered addresses with corresponding Council districts, the module uses the New York Times' Districts API. Unless a working API key is supplied in the module's settings, the module will not function properly. More information about the Districts API is available here: http://developer.nytimes.com/docs/read/districts_api.

A couple of the module's features require custom theme code. These are described below.

1. To enable Tell a Friend emails to be sent as HTML, append the following code to your theme's template.php:

        /**
         * Implementation of hook_webform_mail_headers()
         */
        function nypl_new_webform_mail_headers($node, $submission, $email) {
          $headers = array();
          $tell_a_friend = variable_get('nypl_advocacy_tell_friend_nid', '');
          if ($node->nid == $tell_a_friend) {
            $headers['Content-Type'] = 'text/html; charset=UTF-8; format=flowed; delsp=yes';
          }
          return $headers;
        }

2. To enable Exact Target conversion tracking, add the following code to Webform's default webform-form.tpl.php file, then add the resulting template to your custom theme:

        <?php
        // get Exact Target metadata, if this is the advocacy city council form
        $advocacy_form_nid = variable_get('nypl_advocacy_city_council_nid', NULL);
        if ($form['#node']->nid == $advocacy_form_nid) {
          $et_keys = array('j', 'e', 'l', 'jb', 'u', 'mid');
          foreach ($et_keys as $key) {
            if (isset($_GET[$key]) && (!isset($_COOKIE[$key]) || $_COOKIE[$key] != $_GET[$key])) {
              setcookie($key, $_GET[$key]);
            }
          }
        }
        ?>


CHANGELOG.txt
CHANGELOG.txt

CHANGES for 1.2.0-non-nypl
--------------------------
* commented out NYPL-specific code
* modified README
* modified nypl_advocacy.info

CHANGES for 1.2.0
-----------------
* incorporated advocacy-related code originally implemented in nypl_new theme
** FILE MODIFIED: nypl_advocacy.module
** FILE MODIFIED: nypl_advocacy.install
** FILE ADDED: includes/nypl_advocacy.install.inc
** FILE ADDED: includes/nypl_advocacy.theme.inc
** FILE ADDED: advocacy2012.css
** DIR ADDED: img/socialmedia
** FILES REMOVED: includes/nyc_council_members.inc, includes/webform_nodes.inc (contents moved to nypl_advocacy.install.inc)
* cleaned up code style
** FILE MODIFIED: includes/nypl_advocacy_branch_to_rep.inc
** FILE MODIFIED: includes/nypl_advocacy_generate_toolbox_page.inc
** FILE MODIFIED: includes/nypl_advocacy_helpers.inc
** FILE MODIFIED: includes/nypl_advocacy_image_upload_form.inc
* added documentation
** FILE ADDED: README.txt

CHANGES for 1.1.2
-----------------
* modified data export to pull email data from webform tables, and to pull in data for "generic" location emails
** FILE MODIFIED: nypl_advocacy.module

CHANGES for 1.1
---------------
*  added menu callback for thank you page -> nypl_advocacy_thankyou()
*  added admin form for thank you page text/links -> nypl_advocacy_admin_thankyou_form()
*  added borough field to letters data export -> nypl_advocacy_stats()
*  fixed logic for generating map markers in nypl_advocacy_letters_map()
*  added correct sorting for library branch options in nypl_advocacy_letters_form()
*  added file download for data export -> nypl_advocacy_letters_form()
** FILE MODIFIED: nypl_advocacy.module
*  added helper function _nypl_advocacy_get_badge($sid)
** DIR ADDED:  includes
** FILE ADDED: includes/nypl_advocacy_helpers.inc
** FILE ADDED: includes/nypl_advocacy_branch_to_rep.inc
** FILE ADDED: includes/nypl_advocacy_generate_toolbox_page.inc
** FILE ADDED: includes/nypl_advocacy_image_upload_form.inc
** FILE ADDED: includes/nypl_advocacy_toolbox_form.inc
** DIR ADDED:  img
** DIR ADDED:  img/gmap_markers
** DIR ADDED:  img/web



templates/advocacy_landing.tpl.php
<?php /* $Id: node.tpl.php,v 1.3 2008/06/22 20:49:49 add1sun Exp $ */ 
require_once drupal_get_path('module', 'nypl_advocacy') . '/includes/nypl_advocacy_helpers.inc';
$advocacy_buttons = _get_button_image();
?>
<div id="node-<?php print $node->nid; ?>" class="node<?php print " node-" . $node->type; ?><?php print ($sticky) ? " node-sticky" : ""; ?>">

<?php if (!$page && $title): ?>
<h3 class="title"><a href="<?php print $node_url; ?>" title="<?php print $title .' unaltered!'; ?>"><?php print $title .' unaltered!'; ?></a></h3>
<?php endif; ?>


<?php
if ($node->av_content['nid']) {
	$av_display = theme('nypl_content_av_player', $node->av_content['nid'], 'video', 'simple');
}

if ($av_display && $node->av_content['position'] == 'top') {
	print $av_display;
}

if ($node->image_gallery && $node->image_gallery['position'] == 'top') {
	print $node->image_gallery['content'];
}
?>
<?php echo('hello'); ?>
<div class="content">
  <?php print $field_advocacy_intro[0]['value']; ?>
  <!--  write/donate buttons -->
  <?php print variable_get('nypl_advocacy_write_button_link',''); print $advocacy_buttons['write-button']['tag'];?></a>
  <?php print variable_get('nypl_advocacy_donate_button_link',''); print $advocacy_buttons['donate-button']['tag'];?></a>
  <?php print $content; ?>
  <!-- branch badges -->
  <?php if (variable_get('nypl_advocacy_show_branch_badges', 1) == TRUE){?>
    <p id="advocacy-most-letters-message">Most Letters So Far</p>
    <ul id="advocacy-speakout-badges" class="group">
      <li><?php print _nypl_advocacy_leading_library_badge('bronx'); ?><p class="badge-borough">Bronx</p></li>
      <li><?php print _nypl_advocacy_leading_library_badge('manhattan'); ?><p class="badge-borough">Manhattan</p></li>
      <li><?php print _nypl_advocacy_leading_library_badge('staten_island'); ?><p class="badge-borough">Staten Island</p></li>
    </ul>
  <?php }?>
  <?php print $field_advocacy_txt_bottom[0]['value']; ?>
</div>
  <!-- social media -->
  <?php
    $twitter = variable_get('nypl_advocacy_ty_twitter', '');
    $fb = variable_get('nypl_advocacy_ty_fb', '');
    $pinterest = _nypl_advocacy_replace_pinterest(variable_get('nypl_advocacy_ty_pinterest', ''));
    $google_plus =  variable_get('nypl_advocacy_ty_google_plus', '');
  ?>
  <?php if ($twitter && $fb && $pinterest && $google_plus): ?>
    <ul id="advocacy-speakout-socialmedia" class="group">FOLLOW US ON
      <li class="advocacy-speakout-twitter"><?php print $twitter; ?></li>
      <li class="advocacy-speakout-facebook"><?php print $fb; ?></li>
      <li class="advocacy-speakout-pinterest"><?php print $pinterest; ?></li>
      <li class="advocacy-speakout-google-plus"><?php print $google_plus; ?></li>
    </ul>
  <?php endif; ?>

<?php
if ($av_display && $node->av_content['position'] == 'bottom') {
	print $av_display;
}

if ($node->image_gallery['content'] && $node->image_gallery['position'] == 'bottom') {
	print $node->image_gallery['content'];
}

include('includes/attached_files.php');
 
$path_args = arg();
$context_menu = theme('nypl_content_check_context_menu', $path_args, $node, $current_node); 
print $context_menu;

?>

<div class="links"><?php print $links; ?></div>
</div>

templates/node-advocacy_landing.tpl.php
<?php /* $Id: node.tpl.php,v 1.3 2008/06/22 20:49:49 add1sun Exp $ */ 
require_once drupal_get_path('module', 'nypl_advocacy') . '/includes/nypl_advocacy_helpers.inc';
$advocacy_buttons = _get_button_image();
?>
<div id="node-<?php print $node->nid; ?>" class="node<?php print " node-" . $node->type; ?><?php print ($sticky) ? " node-sticky" : ""; ?>">

<?php if (!$page && $title): ?>
<h3 class="title"><a href="<?php print $node_url; ?>" title="<?php print $title .' unaltered!'; ?>"><?php print $title .' unaltered!'; ?></a></h3>
<?php endif; ?>


<?php
if ($node->av_content['nid']) {
	$av_display = theme('nypl_content_av_player', $node->av_content['nid'], 'video', 'simple');
}

if ($av_display && $node->av_content['position'] == 'top') {
	print $av_display;
}

if ($node->image_gallery && $node->image_gallery['position'] == 'top') {
	print $node->image_gallery['content'];
}
?>

<div class="content">
  <?php print $field_advocacy_intro[0]['value']; ?>
  <!--  write/donate buttons -->
  <?php print variable_get('nypl_advocacy_write_button_link',''); print $advocacy_buttons['write-button']['tag'];?></a>
  <?php print variable_get('nypl_advocacy_donate_button_link',''); print $advocacy_buttons['donate-button']['tag'];?></a>
  <?php print $content; ?>
  <!-- branch badges -->
      <?php if (variable_get('nypl_advocacy_show_branch_badges', 1) == TRUE){?>
        <p id="advocacy-most-letters-message">Most Letters So Far</p>
        <ul id="advocacy-speakout-badges" class="group">
          <li><?php print _nypl_advocacy_leading_library_badge('bronx'); ?><p class="badge-borough">Bronx</p></li>
          <li><?php print _nypl_advocacy_leading_library_badge('manhattan'); ?><p class="badge-borough">Manhattan</p></li>
          <li><?php print _nypl_advocacy_leading_library_badge('staten_island'); ?><p class="badge-borough">Staten Island</p></li>
        </ul>
      <?php }?>
    <?php print $field_advocacy_txt_bottom[0]['value']; ?>
</div>
  <!-- social media -->
  <?php
    $twitter = variable_get('nypl_advocacy_ty_twitter', '');
    $fb = variable_get('nypl_advocacy_ty_fb', '');
    $pinterest = _nypl_advocacy_replace_pinterest(variable_get('nypl_advocacy_ty_pinterest', ''));
    $google_plus =  variable_get('nypl_advocacy_ty_google_plus', '');
  ?>
  <?php if ($twitter && $fb && $pinterest && $google_plus): ?>
    <ul id="advocacy-speakout-socialmedia" class="group">FOLLOW US ON
      <li class="advocacy-speakout-twitter"><?php print $twitter; ?></li>
      <li class="advocacy-speakout-facebook"><?php print $fb; ?></li>
      <li class="advocacy-speakout-pinterest"><?php print $pinterest; ?></li>
      <li class="advocacy-speakout-google-plus"><?php print $google_plus; ?></li>
    </ul>
  <?php endif; ?>

<?php
if ($av_display && $node->av_content['position'] == 'bottom') {
	print $av_display;
}

if ($node->image_gallery['content'] && $node->image_gallery['position'] == 'bottom') {
	print $node->image_gallery['content'];
}

include('includes/attached_files.php');
 
$path_args = arg();
$context_menu = theme('nypl_content_check_context_menu', $path_args, $node, $current_node); 
print $context_menu;

?>

<div class="links"><?php print $links; ?></div>
</div>